felix.go

```diff
@@ -647,51 +655,98 @@ func monitorAndManageShutdown(failureReportChan <-chan string, driverCmd *exec.C
-			globalConfig[key.Name] = value
-		}
+func loadConfigFromDatastore(
+	ctx context.Context, client bapi.Client, hostname string,
+) (globalConfig, hostConfig map[string]string, err error) {
+
+	//log.Info("Waiting for the datastore to be ready")
+	//if kv, err := datastore.Get(ctx, model.ReadyFlagKey{}, ""); err != nil {
+	//	log.WithError(err).Error("Failed to read global datastore 'Ready' flag, will retry...")
+	//	time.Sleep(1 * time.Second)
+	//	continue
+	//} else if kv.Value != true {
+	//	log.Warning("Global datastore 'Ready' flag set to false, waiting...")
+	//	time.Sleep(1 * time.Second)
+	//	continue
+	//}
+
+	g, err := client.Get(ctx, model.ResourceKey{
+		Kind:      apiv2.KindFelixConfiguration,
+		Name:      "default",
+		Namespace: "",
+	}, "")
+	if _, ok := err.(errors2.ErrorResourceDoesNotExist); err != nil && !ok {
+		return
+	}
+	h, err := client.Get(ctx, model.ResourceKey{
+		Kind:      apiv2.KindFelixConfiguration,
+		Name:      "node." + hostname,
+		Namespace: "",
+	}, "")
+	if _, ok := err.(errors2.ErrorResourceDoesNotExist); err != nil && !ok {
+		return
+	}
+	ci, err := client.Get(ctx, model.ResourceKey{
+		Kind:      apiv2.KindClusterInformation,
+		Name:      "node." + hostname,
+		Namespace: "",
+	}, "")
+	if _, ok := err.(errors2.ErrorResourceDoesNotExist); err != nil && !ok {
+		return
+	}
+
+	globalConfig, err = convertV2ConfigToMap("global", g)
+	if err != nil {
+		return
+	}
+	hostConfig, err = convertV2ConfigToMap("per-host", h)
+	if err != nil {
+		return
+	}
 
-		log.Infof("Loading per-host config from datastore; hostname=%v", hostname)
-		kvl, err = datastore.List(ctx,
-			model.HostConfigListOptions{Hostname: hostname}, "")
-		if err != nil {
-			log.WithError(err).Error("Failed to load config from datastore")
-			time.Sleep(1 * time.Second)
-			continue
-		}
-		hostConfig = make(map[string]string)
-		for _, kv := range kvl.KVPairs {
-			key := kv.Key.(model.HostConfigKey)
-			value := kv.Value.(string)
-			hostConfig[key.Name] = value
+	// Merge in the global cluster info config.
+	ciConfig, err := convertV2ConfigToMap("cluster-info", ci)
+	if err != nil {
+		return
+	}
+	for k, v := range ciConfig {
+		globalConfig[k] = v
+	}
+
+	return
+}
+
+// convertV2ConfigToMap converts a v2 datamodel config struct (where each configuration entry
+// is stored in a field of the struct) into a map, as required by our v1-style configuration
+// loader.
+func convertV2ConfigToMap(configType string, v2Config *model.KVPair) (map[string]string, error) {
+	logCxt := log.WithField("type", configType)
+	if v2Config == nil {
+		logCxt.Info("No config of this type")
+		return nil, nil
+	}
+	// Re-use the update processor logic implemented for hte syncer.  We give it a v2 config
+	// object in a KVPair and it uses the annotations defined on it to split it into v1-style
+	// KV pairs.
+	configConvertor := updateprocessors.NewFelixConfigUpdateProcessor()
+	v1kvs, err := configConvertor.Process(v2Config)
```

Comment:
Does this encapsulate the logic for _not_ emitting an update for a particular key/value that is not set in the FelixConfiguration?  (As an aside, I wasn't sure how "not set" is represented for string values.  I got that nil pointers are used for the other cases.)

Reply:
I believe they use an empty string for "not set".  In that case it emits a KVPair with a nil `Value` field.  (The Value field is an `interface{}` so we're talking about a nil interface rather than a  nil pointer inside the interface.)